components/brave_shields/browser/ad_block_regional_service_manager.cc

```diff
@@ -185,47 +200,60 @@ void AdBlockRegionalServiceManager::EnableTag(const std::string& tag,
+void AdBlockRegionalServiceManager::AddResources(const std::string& resources) {
   base::AutoLock lock(regional_services_lock_);
   for (const auto& regional_service : regional_services_) {
     regional_service.second->AddResources(resources);
   }
 }
 
-void AdBlockRegionalServiceManager::EnableFilterList(
-    const std::string& uuid, bool enabled) {
+void AdBlockRegionalServiceManager::EnableFilterList(const std::string& uuid,
+                                                     bool enabled) {
   DCHECK(!uuid.empty());
-  auto catalog_entry = brave_shields::FindAdBlockFilterListByUUID(
-      regional_catalog_, uuid);
+  auto catalog_entry =
+      brave_shields::FindAdBlockFilterListByUUID(regional_catalog_, uuid);
 
   // Enable or disable the specified filter list
-  if (initialized_) {
-    base::AutoLock lock(regional_services_lock_);
-    DCHECK(catalog_entry != regional_catalog_.end());
-    auto it = regional_services_.find(uuid);
-    if (enabled) {
-      DCHECK(it == regional_services_.end());
-      auto regional_service = AdBlockRegionalServiceFactory(
-          *catalog_entry, delegate_,
-          base::BindRepeating(&AdBlockRegionalServiceManager::AddResources,
-                              base::Unretained(this)));
-      regional_service->Start();
-      regional_services_.insert(
-          std::make_pair(uuid, std::move(regional_service)));
-    } else {
-      DCHECK(it != regional_services_.end());
-      it->second->Unregister();
-      regional_services_.erase(it);
-    }
+  base::AutoLock lock(regional_services_lock_);
+  DCHECK(catalog_entry != regional_catalog_.end());
+  auto it = regional_services_.find(uuid);
+  if (enabled) {
+    DCHECK(it == regional_services_.end());
+    auto regional_source_provider =
+        std::make_unique<AdBlockRegionalSourceProvider>(
+            component_update_service_, *catalog_entry);
+    auto regional_service =
+        std::unique_ptr<AdBlockEngine, base::OnTaskRunnerDeleter>(
+            new AdBlockEngine(), base::OnTaskRunnerDeleter(task_runner_));
+    auto observer = std::make_unique<AdBlockService::SourceProviderObserver>(
+        regional_service->AsWeakPtr(), regional_source_provider.get(),
+        resource_provider_, task_runner_);
+    regional_services_.insert(
+        std::make_pair(uuid, std::move(regional_service)));
+    regional_source_providers_.insert(
+        std::make_pair(uuid, std::move(regional_source_provider)));
+    regional_source_observers_.insert(
+        std::make_pair(uuid, std::move(observer)));
+  } else {
+    auto observer = regional_source_observers_.find(uuid);
+    DCHECK(observer != regional_source_observers_.end());
+    regional_source_observers_.erase(observer);
+
+    DCHECK(it != regional_services_.end());
+    regional_services_.erase(it);
+
+    auto it2 = regional_source_providers_.find(uuid);
+    DCHECK(it2 != regional_source_providers_.end());
+    std::move(*it2->second).Delete();
```

Comment:
this is intentional behavior to ensure it can't be used after it is deleted

Reply:
this method was added because the original code unregistered the component when the source provider was destroyed, but that would have caused the component to be uninstalled and reinstalled on every browser restart which is clearly not a good thing. The workaround for now was to add this Delete method that only gets called when we explicitly remove a region